src/hooks/with-async-effect.js

```diff
@@ -0,0 +1,34 @@
+import { useEffect, useState } from 'react'
+
+import { getDeps, isFunction, isPromiseLike, getInternalCtor } from '../utils'
+
+export const withAsyncEffect = (asyncCallback, deps, unmountCallback) => (state, props) => {
+  if (!asyncCallback || !isFunction(asyncCallback)) {
+    throw Error(`withAsyncEffect expects function, got a: ${typeof asyncCallback}`)
+  }
+
+  const [data, setData] = useState({ loading: true, data: null })
+
+  useEffect(() => {
+    const fetch = async () => {
```

Comment:
what for we need this fetch wrapper?
looks like it could be inlined

Reply:
we should wrap async function in useEffect, because there is warning: useEffect function must return a cleanup function or nothing. Promises and useEffect(async () => …) are not supported, but you can call an async function inside an effect.